### PR TITLE
add `StatefulDataLoader` to `knowledge_distillation_single_device` recipe

### PR DESCRIPTION
**Pull request by jxtngx**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#2439 

#### Changelog
What are the changes made in this PR?
Added `StatefulDataLoader` to `knowledge_distillation_single_device` recipe.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [x] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by jxtngx** [comment]
adds remaining recipe mentioned in #2441

File: recipes/knowledge_distillation_single_device.py [hunk 1]
```diff
@@ -582,13 +584,17 @@ def save_checkpoint(self, epoch: int) -> None:
         intermediate_checkpoint = epoch + 1 < self.total_epochs
         # if training is in-progress, checkpoint the optimizer state as well
         if intermediate_checkpoint:
+            dataloader_sd = self._dataloader.state_dict()
```
Comment (joecummings) [inline recipes/knowledge_distillation_single_device.py:587]: Can you actually switch to using the hack from #2431 to advance the iterator? It's (slightly) cleaner than modifying the state dict directly.

**Review by joecummings — COMMENTED** [review]
Only one comment, but this looks great!

File: recipes/knowledge_distillation_single_device.py (hunk 1 above)
Comment (jxtngx) [reply recipes/knowledge_distillation_single_device.py]: resolved in https://github.com/pytorch/torchtune/pull/2442/commits/6d4495aedb1c39d1ae00b2f6c8b2afb7664f286d

Comment (joecummings) [reply recipes/knowledge_distillation_single_device.py]: You still need to advance the iterator after loading with `list(dataloader)`. See #2431

Comment (joecummings) [reply recipes/knowledge_distillation_single_device.py]: [cc6988c](https://github.com/pytorch/torchtune/pull/2442/commits/cc6988cc32c9060bd4fecea0770c3a91be6b73a6) is not right, you'll just want to advance the iterator after loading the state dict into the dataloader as I mentioned above.

Comment (jxtngx) [reply recipes/knowledge_distillation_single_device.py]: ptl @ [32c84c5](https://github.com/pytorch/torchtune/pull/2442/commits/32c84c560882aae017cff0c946c7e7850e19aa61)
